genie-web/src/main/java/com/netflix/genie/web/services/impl/JobResolverServiceImpl.java

```diff
@@ -672,4 +724,214 @@ private int resolveJobMemory(final JobRequest jobRequest, final Command command)
+                    } else {
+                        log.debug("Found {} clusters for {}", clusters.size(), mergedCriterion);
+                        cluster = this.selectClusterUsingClusterSelectors(
+                            Sets.newHashSet(),
+                            clusters,
+                            jobId,
+                            jobRequest
+                        );
+                    }
+                }
+
+                if (cluster != null) {
+                    break;
+                }
+            }
+
+            if (cluster == null) {
+                this.noClusterSelectedCounter.increment();
+                throw new GenieJobResolutionException("No cluster selected given criteria for job " + jobId);
+            }
+
+            log.debug("Selected cluster {} for job {}", cluster.getId(), jobId);
+
+            MetricsUtils.addSuccessTags(tags);
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_ID, cluster.getId()));
+            tags.add(Tag.of(MetricsConstants.TagKeys.CLUSTER_NAME, cluster.getMetadata().getName()));
+            return cluster;
+        } catch (final Throwable t) {
+            MetricsUtils.addFailureTagsWithException(tags, t);
+            if (t instanceof GenieJobResolutionException) {
+                throw (GenieJobResolutionException) t;
+            } else {
+                throw new GenieJobResolutionException(t);
+            }
+        } finally {
+            this.registry
+                .counter(RESOLVE_CLUSTER_CRITERIA_COMBINATION_COUNTER, tags)
+                .increment(criteriaCombinationCount);
+            this.registry
+                .counter(RESOLVE_CLUSTER_QUERY_COUNTER, tags)
+                .increment(queryCount);
+            this.registry
+                .timer(RESOLVE_CLUSTER_TIMER_NAME, tags)
+                .record(System.nanoTime() - start, TimeUnit.NANOSECONDS);
+        }
+    }
+
+    /**
+     * Helper method for merging two criterion.
+     * <p>
+     * This method makes several assumptions:
+     * - If any of these fields: {@literal id, name, version, status} are in both criterion their values must match
+     * or this criterion combination of criteria can't possibly be matched so an {@link IllegalArgumentException}
+     * is thrown
+     * - If only one criterion has any of these fields {@literal id, name, version, status} then that value is present
+     * in the resulting criterion
+     * - Any {@literal tags} present in either criterion are merged into the super set of both sets of tags
+     *
+     * @param one The first {@link Criterion}
+     * @param two The second {@link Criterion}
+     * @return A merged {@link Criterion} that can be used to search the database
+     * @throws IllegalArgumentException If the criteria can't be merged due to the described assumptions
+     */
+    private Criterion mergeCriteria(final Criterion one, final Criterion two) throws IllegalArgumentException {
```

Comment:
I feel like this belong somewhere else, i.e. in the criterion class, maybe as static helper/factory